### ISSUE TITLE
Magic comments that contain frozen_string_literal: true with other directives are marked as violations

### ISSUE THREAD

**Issue by backus** [bbatsov/rubocop#3285]
Rubocop correctly approves of files that have a comment like `# frozen_string_literal: true`. You can also combine multiple magic directives by writing something like `# -*- encoding: ASCII-8BIT; frozen_string_literal: true -*-`. This doesn't work for me right now.

Example / steps to reproduce:

``` ruby
$ cat example.rb
# -*- encoding: ASCII-8BIT; frozen_string_literal: true -*-

foo = 'bar'
puts foo.frozen?
```

``` sh
$ ruby example.rb
true
```

``` sh
$ rubocop example.rb
Inspecting 1 file
C

Offenses:

example.rb:1:1: C: Missing frozen string literal comment.
# -*- encoding: ASCII-8BIT; frozen_string_literal: true -*-
^

1 file inspected, 1 offense detected
```

```
$ rubocop -V
0.41.1 (using Parser 2.3.1.2, running on ruby 2.3.1 x86_64-darwin15)
```

**Comment by backus** [comment]
It is also valid to split these magic comments into multiple lines:

```sh
$ cat example.rb
# encoding: ASCII-8BIT
# frozen_string_literal: true

foo = 'bar'
puts "encoding: #{foo.encoding}"
puts "frozen:   #{foo.frozen?}"
$ ruby example.rb
encoding: ASCII-8BIT
frozen:   true
```

Maybe the better solution here would be for the ruby style guide to take a stance on these options. I think the preferred style should be one magic comment per line. Then rubocop could enforce this rule instead and the frozen string magic comment cop wouldn't have to do more complicated parsing

**Comment by Drenmi** [comment]
Hm. I agree that the style guide could take sides here, but I think even if it does, it would be nice to have the cop tell you to split it into separate lines rather than (incorrectly) saying it's missing. That could also open up for auto-correct.

@backus: Do you have a link to where this is documented? I had honestly never seen it before.

**Comment by backus** [comment]
I learned about it through Parser's codebase

I can't find any explicit documentation. Best links I can find just point me to the parsing logic for irb and ruby itself:

- [IRB's magic-file.rb](https://github.com/ruby/ruby/blob/trunk/lib/irb/magic-file.rb)
- [MRI's parse.y](https://github.com/ruby/ruby/blob/ruby_2_2/parse.y#L6980)

Here is how it is used in Parser: https://github.com/whitequark/parser/blob/master/Rakefile#L39

**Comment by backus** [comment]
@Drenmi this is the only other resource I can really point you too: documented via assertions in [rubyspec](https://github.com/ruby/spec/blob/master/language/magic_comment_spec.rb). Ruby's magic comments try to support both emacs and vim style. It doesn't include it in the example but I believe the example I gave works because emacs style magic comments are separated with semicolons (https://www.gnu.org/software/emacs/manual/html_node/emacs/Specify-Coding.html).

**Comment by Drenmi** [comment]
Awesome @backus! Thanks for the research effort.

**Comment by rrosenblum** [comment]
Interesting, thanks for the info @backus and @Drenmi. Even if the style guide takes a stances on this, it is still something that RuboCop should be able to parse in multiple formats. That does unfortunately add complexity to RuboCop. As far as I can tell, there are [3 possible magic directives](http://idiosyncratic-ruby.com/58-magic-instructions.html) that can be specified in Ruby `encoding`, `frozen_string_literals`, and `warn_indent`. RuboCop has a cop for `frozen_string_literal` and `encoding`. Both of these cops would to be able parse magic directives in any form or order.

**Comment by bbatsov** [comment]
I think we should have a style guide entry favouring one directive per line and the use of the Ruby comment syntax, but that being said - obviously it'd be better if we processed properly all valid scenarios.

**Comment by backus** [comment]
On the topic of magic comments: it would be nice if the style guide also included a rule that said that magic comments should be separated from source code by a blank newline. Example:

**Good*

```ruby
# frozen_string_literal: true

class Foo
end
```

**Bad**

```ruby
# frozen_string_literal: true
class Foo
end
```

**Comment by backus** [comment]
> As far as I can tell, there are 3 possible magic directives that can be specified in Ruby encoding, frozen_string_literals, and warn_indent. RuboCop has a cop for frozen_string_literal and encoding. Both of these cops would to be able parse magic directives in any form or order.

@rrosenblum Esoteric ruby trivia: `warn_past_scope` is also a magic comment but you'll probably never see it because it needs a special compiler flag. I'm looking into this issue now so figured I would add this factoid :smile:

**Comment by backus** [comment]
PR here to fix this issue https://github.com/bbatsov/rubocop/pull/3879
I also opened a style guide PR for the things we discussed here https://github.com/bbatsov/ruby-style-guide/pull/616

**Comment by rrosenblum** [comment]
> Esoteric ruby trivia: `warn_past_scope` is also a magic comment but you'll probably never see it because it needs a special compiler flag.

You are right, I did not know about this one. I have not seen any information on it before now.

**Comment by backus** [comment]
@bbatsov @rrosenblum updated. I updated `Style/Encoding` as well. As far as I can tell there was two test cases that were testing invalid encoding comments (`# vim:fileencoding=utf-8`). See the commit message for more context

**Comment by aried3r** [comment]
> On the topic of magic comments: it would be nice if the style guide also included a rule that said that magic comments should be separated from source code by a blank newline.

Hey, sorry for bringing this up again, but is this just a style decision or is there a reason for this to be found in the Ruby parser?

It's a also a bit weird to see in the [Ruby styleguide](https://github.com/bbatsov/ruby-style-guide/blob/master/README.md#magic-comments), first it's ok to have it, then on the bottom it isn't anymore.
